keras_retinanet/models/retinanet.py

```diff
@@ -193,28 +309,72 @@ def retinanet(
-    anchors = __build_anchors(anchor_parameters, features)
+    pyramids = __build_pyramid(submodels, features)
+    anchors   = __build_anchors(anchor_parameters, features)
 
-    return keras.models.Model(inputs=inputs, outputs=[anchors] + pyramid, name=name)
+    # concatenate outputs to one list
+    outputs = [anchors] + pyramids
+    if output_fpn:
+        outputs += features
 
+    return keras.models.Model(inputs=inputs, outputs=outputs, name=name)
 
-def retinanet_bbox(inputs, num_classes, nms=True, name='retinanet-bbox', *args, **kwargs):
-    model = retinanet(inputs=inputs, num_classes=num_classes, *args, **kwargs)
+
+def retinanet_bbox(
+    inputs,
+    num_classes,
+    nms=True,
+    name='retinanet-bbox',
+    output_fpn=False,
+    **kwargs
+):
+    """ Construct a RetinaNet model on top of a backbone and adds convenience functions to output detections directly.
+
+    This model uses the minimum retinanet model and appends a few layers to compute detections within the graph.
+    These layers include applying the regression values to the anchors and performing NMS.
+
+    Args
+        inputs      : keras.layers.Input (or list of) for the input to the model.
+        num_classes : Number of classes to classify.
+        output_fpn  : If True, the last 5 outputs before detections are P3...P7.
+        name        : Name of the model.
+        *kwargs     : Additional kwargs to pass to the minimal retinanet model.
+
+    Returns
+        Model with inputs as input and as output the output of each submodel for each pyramid level and the detections.
+
+        The order is as defined in submodels. Using default values the output is:
+        ```
+        [
+            regression, classification, detections
+        ]
+        ```
+    """
+    model = retinanet(inputs=inputs, num_classes=num_classes, output_fpn=output_fpn, **kwargs)
 
     # we expect the anchors, regression and classification values as first output
     anchors        = model.outputs[0]
     regression     = model.outputs[1]
     classification = model.outputs[2]
 
+    if output_fpn:
+        # "other" can be any additional output from custom submodels, by default this will be []
+        other = model.outputs[3:-5]
+        fpn = model.outputs[-5:]
+    else:
+        other = model.outputs[3:]
+
     # apply predicted regression to anchors
     boxes = layers.RegressBoxes(name='boxes')([anchors, regression])
 
     # additionally apply non maximum suppression
     if nms:
         detections = layers.NonMaximumSuppression(name='nms')([boxes, classification] + model.outputs[3:])
     else:
-        detections = keras.layers.Concatenate(axis=2)([boxes, classification] + model.outputs[3:])
+        detections = keras.layers.Concatenate(axis=2, name='detections')([boxes, classification] + model.outputs[3:])
+
+    # construct list of outputs
+    outputs = [regression, classification] + other + (fpn if output_fpn else []) + [detections]
```

Comment:
This is the only place I see `other` and `fpn` being used, and they are being concatenated. Do we really need to distinguish `output_fpn` and `not output_fpn` then? It looks like `other = model.outputs[3:]` would suffice in both cases (in which case I would also replace all occurrences of `model.outputs[3:]` with `other`). If so, output_fpn can also be caught by **kwargs.

If `output_fpn` can't be removed, how about just setting `fpn = []` in the `not output_fpn` case above, and simplify this to `other + fpn`?

Reply:
Ah actually that was a boo-boo which I already fixed, but didn't push yet. `other` should've been used instead of `model.outputs[3:]` (which it does now). The thing is that the last 5 elements are the FPN features if `output_fpn` is used, so `model.outputs[3:]` wouldn't work in every case.